lib/remote_node.go

```diff
@@ -0,0 +1,744 @@
+}
+
+// sendVerack constructs and sends a verack message to the peer.
+func (rn *RemoteNode) sendVerack() error {
+	verackMsg, err := rn.newVerackMessage()
+	if err != nil {
+		return err
+	}
+
+	if err := rn.sendMessage(verackMsg); err != nil {
+		return errors.Wrapf(err, "RemoteNode.SendVerack: Problem sending verack message to peer (id= %d): %v", rn.id, err)
+	}
+	return nil
+}
+
+// newVerackMessage constructs a verack message to be sent to the peer.
+func (rn *RemoteNode) newVerackMessage() (*MsgDeSoVerack, error) {
+	verack := NewMessage(MsgTypeVerack).(*MsgDeSoVerack)
+	vMeta := rn.handshakeMetadata
+
+	switch vMeta.negotiatedProtocolVersion {
+	case ProtocolVersion0, ProtocolVersion1:
+		// For protocol versions 0 and 1, we just send back the nonce we received from the peer in the version message.
+		verack.Version = VerackVersion0
+		verack.NonceReceived = vMeta.versionNonceReceived
+	case ProtocolVersion2:
+		// For protocol version 2, we need to send the nonce we received from the peer in their version message.
+		// We also need to send our own nonce, which we generate for our version message. In addition, we need to
+		// send a current timestamp (in microseconds). We then sign the tuple of (nonceReceived, nonceSent, tstampMicro)
+		// using our validator BLS key, and send the signature along with our public key.
+		var err error
+		verack.Version = VerackVersion1
+		verack.NonceReceived = vMeta.versionNonceReceived
+		verack.NonceSent = vMeta.versionNonceSent
+		tstampMicro := uint64(time.Now().UnixMicro())
+		verack.TstampMicro = tstampMicro
+		// If the RemoteNode is not a validator, then we don't need to sign the verack message.
+		if !rn.nodeServices.HasService(SFPosValidator) {
+			break
+		}
+		verack.PublicKey = rn.keystore.GetSigner().GetPublicKey()
+		verack.Signature, err = rn.keystore.GetSigner().SignPoSValidatorHandshake(verack.NonceSent, verack.NonceReceived, tstampMicro)
+		if err != nil {
+			return nil, fmt.Errorf("RemoteNode.newVerackMessage: Problem signing verack message: %v", err)
+		}
+	}
+	return verack, nil
+}
+
+// HandleVerackMessage handles a verack message received from the peer.
+func (rn *RemoteNode) HandleVerackMessage(vrkMsg *MsgDeSoVerack) error {
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	if rn.connectionStatus != RemoteNodeStatus_VerackSent {
+		return fmt.Errorf("RemoteNode.HandleVerackMessage: Requesting disconnect for id: (%v) "+
+			"verack received while in state: %v", rn.id, rn.connectionStatus)
+	}
+
+	if rn.verackTimeExpected != nil && rn.verackTimeExpected.Before(time.Now()) {
+		return fmt.Errorf("RemoteNode.HandleVerackMessage: Requesting disconnect for id: (%v) "+
+			"verack timeout. Time expected: %v, now: %v", rn.id, rn.verackTimeExpected.UnixMicro(), time.Now().UnixMicro())
+	}
+
+	var err error
+	vMeta := rn.handshakeMetadata
+	switch vMeta.negotiatedProtocolVersion {
+	case ProtocolVersion0, ProtocolVersion1:
+		err = rn.validateVerackPoW(vrkMsg)
+	case ProtocolVersion2:
```

Comment:
I think this will still work, you just need to be more lenient with regard to version negotiation. In particular, you need to allow negotiated version to be ProtocolVersion1 even if you're on ProtocolVersion2.